lib/tree.rb

```diff
@@ -16,51 +16,137 @@ def initialize
     @root = nil
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
-  def add(key, value)
-    raise NotImplementedError
+  # Time Complexity: O(logn)
+  # Space Complexity: O(logn)
+  def add(key, value = nil)
+    unless root
+      @root = TreeNode.new(key, value)
+    else
+      add_helper(@root, key, value)
+    end
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  def add_helper(current, key, value)
+    return TreeNode.new(key, value) unless current
+
+    if key < current.key
+      current.left = add_helper(current.left, key, value)
+    else
+      current.right = add_helper(current.right, key, value)
+    end
+
+    return current
+
+  end
+
+  # Time Complexity: O(logn)
+  # Space Complexity: O(logn)
   def find(key)
```

Comment:
👍 The time/complexities are correct if the tree is balanced and O(n) if unbalanced.